[inductor] Fix an aliased output bug

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115373

Summary: for https://github.com/pytorch/pytorch/issues/97083, when

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

desertfire committed:
[inductor] Fix an alias-related bug

Summary: for https://github.com/pytorch/pytorch/issues/97083, when

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Fix an aliased output bug"


Summary: for https://github.com/pytorch/pytorch/issues/97083, when

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Fix an aliased output bug"


Summary: for https://github.com/pytorch/pytorch/issues/97083, when

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented on torch/_inductor/graph.py:
```diff
@@ -840,6 +840,14 @@ def debug(msg):
+                and isinstance(result, TensorBox)
+                and isinstance(result.data, ir.BaseView)
+            ):
+                # Realize so that outputs are correctly alised
```
would you say a little bit more about how this fixes things ? not sure i follow

desertfire replied:
In the new unit test, there is a graph break because of the data-dependent control flow, and here are the last few statements in the first graph, 
```
slice_5: "i64[1]" = torch.ops.aten.slice.Tensor(slice_scatter, 0, 0, 1)
squeeze_1: "i64[]" = torch.ops.aten.squeeze.default(slice_5);  slice_5 = None
slice_6: "i64[1]" = torch.ops.aten.slice.Tensor(slice_scatter, 0, 0, 1);  slice_scatter = None
full_default: "b8[]" = torch.ops.aten.full.default([], True, dtype = torch.bool, layout = torch.strided, device = device(type='cuda', index=0), pin_memory = False)
return (squeeze_1, slice_6, full_default)
```
, where `squeeze_1` and `slice_6` are aliased. Before this change, both of them will become unrealized `Pointwise` and two different buffers. Then in the second graph, updating to `slice_6` will not be reflected in `squeeze_1`.

desertfire commented:
@pytorchbot merge

atalman commented:
@pytorchbot revert -m 'Sorry for reverting your change it broke inductor tests' -c nosignal

desertfire committed:
Update on "[inductor] Fix an aliased output bug"


Summary: for https://github.com/pytorch/pytorch/issues/97083, when

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge
